crates/proto/src/rr/rdata/cert.rs

```diff
@@ -0,0 +1,762 @@
+///    for transaction security mechanisms (SIG(0) and TSIG), and some for
+///    both.  Those usable for zone signing may appear in DNSKEY, RRSIG, and
+///    DS RRs.  Those usable for transaction security would be present in
+///    SIG(0) and KEY RRs, as described in [RFC2931].
+///
+///                                 Zone
+///    Value Algorithm [Mnemonic]  Signing  References   Status
+///    ----- -------------------- --------- ----------  ---------
+///      0   reserved
+///      1   RSA/MD5 [RSAMD5]         n      [RFC2537]  NOT RECOMMENDED
+///      2   Diffie-Hellman [DH]      n      [RFC2539]   -
+///      3   DSA/SHA-1 [DSA]          y      [RFC2536]  OPTIONAL
+///      4   Elliptic Curve [ECC]              TBA       -
+///      5   RSA/SHA-1 [RSASHA1]      y      [RFC3110]  MANDATORY
+///    252   Indirect [INDIRECT]      n                  -
+///    253   Private [PRIVATEDNS]     y      see below  OPTIONAL
+///    254   Private [PRIVATEOID]     y      see below  OPTIONAL
+///    255   reserved
+///
+///    6 - 251  Available for assignment by IETF Standards Action.
+///
+/// (RFC Required) Domain Name System Security (DNSSEC) Algorithm Numbers
+/// Created: 2003-11-03, Last Updated: 2024-04-16
+/// https://www.iana.org/assignments/dns-sec-alg-numbers/dns-sec-alg-numbers.txt
+///
+///                                                              Zone
+///     Value  Algorithm [Mnemonic]                            Signing    References
+///     -----  --------------------                           ---------   ----------
+///       6    DSA-NSEC3-SHA1 [DSA-NSEC3-SHA1]                    Y       [RFC5155][proposed standard]
+///       7    RSASHA1-NSEC3-SHA1 [RSASHA1-NSEC3-SHA1]            Y       [RFC5155][proposed standard]
+///       8    RSA/SHA-256 [RSASHA256]                            Y       [RFC5702][proposed standard]
+///       9    reserved
+///      10    RSA/SHA-512 [RSASHA512]                            Y       [RFC5702][proposed standard]
+///      11    reserved
+///      12    GOST R 34.10-2001 [ECC-GOST]                       Y       [RFC5933][proposed standard]
+///      13    ECDSA Curve P-256 with SHA-256 [ECDSAP256SHA256]   Y       [RFC6605][proposed standard]
+///      14    ECDSA Curve P-384 with SHA-384 [ECDSAP384SHA384]   Y       [RFC6605][proposed standard]
+///      15    Ed25519 [ED25519]                                  Y       [RFC8080][proposed standard]
+///      16    Ed448 [ED448]                                      Y       [RFC8080][proposed standard]
+///      17    SM2 signing with SM3 hashing [SM2SM3]              Y       [RFC-cuiling-dnsop-sm2-alg-15][informational]
+///   18-22    Unassigned
+///      23    GOST R 34.10-2012 [ECC-GOST12]                     Y       [RFC9558][informational]
+///  24-122    Unassigned
+/// 123-251    reserved
+///
+/// ```
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum Algorithm {
```

Comment:
I do realize I'm (very) late to the party, but why was this added? This is the same data type as https://github.com/hickory-dns/hickory-dns/blob/7887002d1198ed9faaaae531e0f9815c10589f7a/crates/proto/src/rr/dnssec/algorithm.rs#L105?

Reply:
Because the original reporter was unaware, and the reviewers didn't catch it... I started on a PR to clean it up.

Reply:
Thanks :)

Reply:
> I do realize I'm (very) late to the party, but why was this added? This is the same data type as 

Because it isn't really the same. While the numbers are the same and come from the same ultimate source, the valid variants for each record type are different.  From [IANA DNS Security Algorithm Numbers](https://www.iana.org/assignments/dns-sec-alg-numbers/dns-sec-alg-numbers.txt) document:

> All algorithm numbers in this registry may be used in CERT RRs. Zone
 signing (DNSSEC) and transaction security mechanisms (SIG(0) and TSIG)
 make use of particular subsets of these algorithms. Only algorithms
 usable for zone signing may appear in DNSKEY, RRSIG, and DS RRs.
 Only those usable for SIG(0) and TSIG may appear in SIG and KEY RRs.

There are also diverging standards for what algorithm 0 signifies.  From RFC 8078:

> [4](https://www.rfc-editor.org/rfc/rfc8078#section-4).  DNSSEC Delete Algorithm
>
>   This document defines the previously reserved DNS Security Algorithm
   Number of value 0 in the context of CDS and CDNSKEY records to mean
   that the entire DS RRset at the parent must be removed.  The value 0
   remains reserved for the DS and DNSKEY records.
>
>  No DNSSEC validator can treat algorithm 0 as a valid signature
   algorithm.  If a validator sees a DNSKEY or DS record with this
   algorithm value, it must treat it as unknown.  Accordingly, the zone
   is treated as unsigned unless there are other algorithms present.  In
   general, the value 0 should never be used in the context of DNSKEY
   and DS records.
>
>   The CERT record [[RFC4398](https://www.rfc-editor.org/rfc/rfc4398)] defines the value 0 similarly to mean the algorithm in the CERT record is not defined in DNSSEC.

Ignoring the issue with algorithm zero, you could perhaps argue the algorithm list should be in a record-agnostic table. I do not think it is appropriate to bolt a bunch of variants onto the DNSSEC algorithm enum that are not relevant to DNSSEC, or impose DNSSEC-specific constraints on the CERT type.

Reply:
Would it make sense to have a separate enum that lives outside of both the `dnssec` module and the `cert` module that both of these refer to?

Reply:
when I was looking/working between these the possibility they could diverge was easier to protect by ensuring complete coverage that meets cert only as there is algorithms that are implemented but not part of an RFC yet

Reply:
> Would it make sense to have a separate enum that lives outside of both the dnssec module and the cert module that both of these refer to?

I'm not philosophically opposed to the idea, but my main concerns in reviewing such a hypothetical PR would be:

1. Does making this change make record implementations harder to write and maintain by creating a 'worst-of-both-worlds' scenario where there is a central Algorithms enum, but also further constrained RecordTypeAlgorithm enums?

2. Does making this change make the implementations harder to use by turning scenarios that could be caught at compile-time into an error class that can only be detected at run time?

Reply:
Makes sense. Thanks for the insight!